### PR TITLE
events endpoint: fix panic and race condition

### PR DESCRIPTION
**Pull request by vrothberg**
Fix a potential panic in the events endpoint when parsing the filters
parameter.  Values of the filters map might be empty, so we need to
account for that instead of uncondtitionally accessing the first item.

Also apply a similar fix for race conditions as done in commit f4a2d25c0fca:

	Fix a race that could cause read errors to be masked.  Masking
	such errors is likely to report red herrings since users don't
	see that reading failed for some reasons but that a given event
	could not be found.

Fixes: #6899
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@baude @mheon PTAL  before of the spaghetti

File: pkg/api/handlers/compat/events.go [hunk 1]
```diff
@@ -46,18 +52,21 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
 		fromStart = true
 	}
 
+	// Note: the code looks a bit like spaghetti as a couple of things
+	// happen concurrently.  The events API does not have an error channel,
+	// so reading from the events channel and checking for an error must
+	// be handled us here.  We may also stream the events, which adds yet
```
Comment (rhatdan) [inline pkg/api/handlers/compat/events.go:58]: nit `be handled here...`

**Comment by rhatdan** [comment]
LGTM, one nit.

File: pkg/api/handlers/compat/events.go (hunk 1 above)
Comment (vrothberg) [reply pkg/api/handlers/compat/events.go]: Fixed, thanks!

File: pkg/api/handlers/compat/events.go [hunk 2]
```diff
@@ -68,24 +77,52 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
...
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-	if flusher, ok := w.(http.Flusher); ok {
-		flusher.Flush()
-	}
+	// Write the events from the channel.
+	go func() {
+		first := true
+		var coder *jsoniter.Encoder
+		json := jsoniter.ConfigCompatibleWithStandardLibrary
```
Comment (mheon) [inline pkg/api/handlers/compat/events.go:84]: Why do we need to redefine this? This should be set for the entire package

File: pkg/api/handlers/compat/events.go [hunk 3]
```diff
@@ -68,24 +77,52 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
...
+			w.WriteHeader(http.StatusOK)
+			if flusher, ok := w.(http.Flusher); ok {
+				flusher.Flush()
+			}
 
-	for event := range eventChannel {
-		e := entities.ConvertToEntitiesEvent(*event)
-		if err := coder.Encode(e); err != nil {
-			logrus.Errorf("unable to write json: %q", err)
+			coder = json.NewEncoder(w)
```
Comment (mheon) [inline pkg/api/handlers/compat/events.go:95]: Why are these in writeHeader? They seem like they ought to be part of initializing variables above

File: pkg/api/handlers/compat/events.go [hunk 4]
```diff
@@ -68,24 +77,52 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
...
+			coder.SetEscapeHTML(true)
 		}
-		if flusher, ok := w.(http.Flusher); ok {
-			flusher.Flush()
+
+		for event := range eventChannel {
```
Comment (mheon) [inline pkg/api/handlers/compat/events.go:99]: Do we have a guarantee this terminates in non-error conditions? I don't think that the channel is closed automatically when events run out, so we could block forever here.

File: pkg/api/handlers/compat/events.go (hunk 3 above)
Comment (vrothberg) [reply pkg/api/handlers/compat/events.go:95]: Let me rewrite a bit more. The code is hard to parse as is.

File: pkg/api/handlers/compat/events.go (hunk 2 above)
Comment (vrothberg) [reply pkg/api/handlers/compat/events.go:84]: No idea. I will add a TODO as I consider it to be out of context of this PR. I don't want to change to much in a fix.

**Comment by vrothberg** [comment]
@mheon, I did another round of refactoring and believe we're at a fairly elegant place now

File: pkg/api/handlers/compat/events.go (hunk 2 above)
Comment (mheon) [reply pkg/api/handlers/compat/events.go]: I was going to say that I was fairly certain it was already set, but it looks like I'm wrong. Which is actually quite concerning, perf-wise - we have a super-efficient JSON encoder/decoder but we're just not using it in the API, which deals heavily with JSON.

**Comment by TomSweeneyRedHat** [comment]
@vrothberg tests are falling down all over.

**Comment by vrothberg** [comment]
Almost there. I'll take another look. Want to get this in eventually

**Comment by rhatdan** [comment]
@giuseppe @mheon @baude @jwhonce @ashley-cui @QiWang19 @TomSweeneyRedHat PTAL

**Comment by vrothberg** [comment]
@mheon we need that in v2.0

**Comment by vrothberg** [comment]
Please wait before merging. I will add another commit to fix https://github.com/containers/podman/issues/6899.

**Comment by vrothberg** [comment]
> Please wait before merging. I will add another commit to fix #6899.

Done. @rhatdan can you take another look?

File: pkg/api/handlers/compat/events.go [hunk 5]
```diff
@@ -15,77 +16,127 @@ import (
...
+	if err := json.Unmarshal(raw, &filters); err == nil {
+		for filterKey, filterSlice := range filters {
+			for _, filterValue := range filterSlice {
+				libpodFilters = append(libpodFilters, fmt.Sprintf("%s=%s", filterKey, filterValue))
+			}
+		}
+		return libpodFilters
+	}
+
+	return nil
```
Comment (rhatdan) [inline pkg/api/handlers/compat/events.go:56]: This function should return an err correct?  If the second unmarshal fails, we got bad data

Comment (vrothberg) [reply pkg/api/handlers/compat/events.go:56]: Good catch, yes! I'll wrap it up and repush in a bit.

File: pkg/api/handlers/compat/events.go [hunk 6]
```diff
@@ -15,77 +16,132 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+// filtersFromRequests extracts the "filters" parameter from the specified
+// http.Request.  The paramater can either be a `map[string][]string` as done
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/compat/events.go:20]: nit of a nit, only change if you've other changes inbound.
```suggestion
// http.Request.  The parameter can either be a `map[string][]string` as done
```

**Comment by TomSweeneyRedHat** [comment]
LGTM but I want a @mheon head nod before merging.
